ReferencePluginI/ControlI.cs

```diff
@@ -0,0 +1,433 @@
+				else if (token is IUSFMTextToken textToken)
+				{
+					TextToken newToken = new TextToken(textToken);
+					m_Tokens.Add(newToken);
+					labelTextBox.Text = "Text";
+					dataTextBox.Text = textToken.Text;
+					dataTextBox.ReadOnly = false;
+					dataTextBox.Multiline = true;
+					dataTextBox.ScrollBars = ScrollBars.Both;
+					dataTextBox.TextChanged += new System.EventHandler(this.ScriptureTextChanged);
+				}
+				else if (token is IUSFMAttributeToken)
+				{
+					m_Tokens.Add(token);
+					labelTextBox.Text = "Attribute";
+					dataTextBox.Text = token.ToString();
+				}
+				else
+				{
+					m_Tokens.Add(token);
+					labelTextBox.Text = "Other";
+					dataTextBox.Text = token.ToString();
+				}
+
+				tableLayoutPanel.Controls.Add(labelTextBox);
+				tableLayoutPanel.Controls.Add(dataTextBox);
+			}
+			tableLayoutPanel.ResumeLayout();
+
+		}
+
+		public void PutScripture(object sender, EventArgs eArgs)
+		{
+			PutScripture();
+		}
+
+		private void PutScripture()
+		{
+			if (m_Project == null)
+			{
+				MessageBox.Show("No project selected");
+				return;
+			}
+
+			if (m_WriteLock == null)
+			{
+				MessageBox.Show("'Get' before trying to 'Put'");
+				return;
+			}
+
+			if (false == changedCheckBox.Checked)
+			{
+				MessageBox.Show("Data not changed; won't Put it");
+				return;
+			}
+
+			bool putSucceeded = false;
+			if (UsfmRadioButton.Checked)
+			{
+				string text = chapterText.Text;
+				try
+				{
+					m_Project.PutUSFM(m_WriteLock, text, m_Reference.BookNum);
+					putSucceeded = true;
+				}
+				catch (Exception ex)
+				{
+					MessageBox.Show($"Exception caught:\n{ex.Message}");
+				}
+			}
+			else if (UsfmTokensRadioButton.Checked)
+			{
+				UpdateModifiedTokens();
+				try
+				{
+					m_Project.PutUSFMTokens(m_WriteLock, m_Tokens, m_Reference.BookNum);
+					putSucceeded = true;
+				}
+				catch (Exception ex)
+				{
+					MessageBox.Show($"Exception caught:\n{ex.Message}");
+				}
+			}
+			else if (UsxRadioButton.Checked)
+			{
+				string text = chapterText.Text;
+				try
+				{
+					m_Project.PutUSX(m_WriteLock, text);
+					putSucceeded = true;
+				}
+				catch (Exception ex)
+				{
+					MessageBox.Show($"Exception caught:\n{ex.Message}");
+				}
+			}
+
+			if (putSucceeded)
+			{
+				changedCheckBox.Checked = false;
+				m_WriteLock.SendNotifications();
+			}
+		}
+
+		private void ReferenceChanged(IPluginChildWindow sender, IVerseRef oldReference, IVerseRef newReference)
+		{
+			if ((m_Reference.BookNum != newReference.BookNum)
+				|| (m_Reference.ChapterNum != newReference.ChapterNum))
+			{
+				PromptSaveAndUnlock();
+				m_Reference = newReference;
+				chapterText.Text = "Click 'Get Chapter'";
+				chapterText.BringToFront();
+				bookName.Text = m_Reference.BookCode;
+				chapterNumber.Text = m_Reference.ChapterNum.ToString();
+				changedCheckBox.Checked = false;
+			}
+		}
+
+
+		private void UpdateModifiedTokens()
+		{
+			int row = 0;
+			foreach (var token in m_Tokens)
```

Comment:
It would be more efficient to just use a for loop since you need the index anyway.